File: flutter_tapresearch-1.4.0/ios/flutter_tapresearch/Sources/flutter_tapresearch/FlutterTapresearchPlugin.swift
import Flutter
import UIKit
import TapResearchSDK

/// Flutter bridge for the TapResearch iOS SDK 3.x.
///
/// The 3.x SDK requires an api token and a user identifier at initialization,
/// uses string placement tags, and delivers rewards as a batched array via the
/// `TapResearchSDKDelegate`. Content show/dismiss events arrive via the
/// `TapResearchContentDelegate` passed to `showContent`.
public class FlutterTapresearchPlugin: NSObject, FlutterPlugin,
    TapResearchSDKDelegate, TapResearchContentDelegate {

  private static var channel: FlutterMethodChannel?
  private var sdkReady = false

  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(
      name: "flutter_tapresearch", binaryMessenger: registrar.messenger())
    let instance = FlutterTapresearchPlugin()
    FlutterTapresearchPlugin.channel = channel
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    let args = call.arguments as? [String: Any] ?? [:]
    switch call.method {
    case "initialize":
      guard let token = args["apiToken"] as? String, !token.isEmpty else {
        result(FlutterError(code: "no_api_token",
                            message: "A null/empty api token was provided", details: nil))
        return
      }
      guard let userId = args["userIdentifier"] as? String, !userId.isEmpty else {
        result(FlutterError(code: "no_user_id",
                            message: "A null/empty user identifier was provided", details: nil))
        return
      }
      TapResearch.initialize(withAPIToken: token, userIdentifier: userId, sdkDelegate: self) {
        [weak self] error in
        if let error = error {
          self?.invokeError(error)
        }
      }
      result(nil)

    case "isReady":
      result(sdkReady)

    case "canShowContentForPlacement":
      let tag = args["placementTag"] as? String ?? ""
      let canShow = TapResearch.canShowContent(forPlacement: tag) { [weak self] error in
        if let error = error { self?.invokeError(error) }
      }
      result(canShow)

    case "showContentForPlacement":
      let tag = args["placementTag"] as? String ?? ""
      let custom = args["customParameters"] as? [String: Any] ?? [:]
      TapResearch.showContent(forPlacement: tag, delegate: self, customParameters: custom) {
        [weak self] error in
        if let error = error { self?.invokeError(error) }
      }
      result(nil)

    case "setUserIdentifier":
      let userId = args["userIdentifier"] as? String ?? ""
      TapResearch.setUserIdentifier(userId)
      result(nil)

    case "sendUserAttributes":
      let attrs = args["attributes"] as? [String: Any] ?? [:]
      _ = TapResearch.sendUserAttributes(attributes: attrs, clearPreviousAttributes: false)
      result(nil)

    default:
      result(FlutterMethodNotImplemented)
    }
  }

  // MARK: - TapResearchSDKDelegate

  public func onTapResearchDidReceiveRewards(_ rewards: [TRReward]) {
    let mapped = rewards.map { rewardMap($0) }
    FlutterTapresearchPlugin.channel?.invokeMethod("onRewards", arguments: mapped)
  }

  public func onTapResearchDidError(_ error: NSError) {
    invokeError(error)
  }

  public func onTapResearchSdkReady() {
    sdkReady = true
    FlutterTapresearchPlugin.channel?.invokeMethod("onSdkReady", arguments: nil)
  }

  public func onTapResearchQuickQuestionResponse(_ qqPayload: TRQQDataPayload) {
    // Quick Questions are not yet surfaced to Dart. Hook added for forward
    // compatibility; wire up an "onQuickQuestionResponse" call here when needed.
  }

  // MARK: - TapResearchContentDelegate

  public func onTapResearchContentShown(forPlacement placementTag: String) {
    FlutterTapresearchPlugin.channel?.invokeMethod("onContentShown", arguments: placementTag)
  }

  public func onTapResearchContentDismissed(forPlacement placementTag: String) {
    FlutterTapresearchPlugin.channel?.invokeMethod("onContentDismissed", arguments: placementTag)
  }

  // MARK: - Helpers

  private func rewardMap(_ reward: TRReward) -> [String: Any] {
    return [
      "transactionIdentifier": reward.transactionIdentifier,
      "placementIdentifier": reward.placementIdentifier,
      "placementTag": reward.placementTag,
      "currencyName": reward.currencyName,
      "rewardAmount": reward.rewardAmount,
      "payoutEventType": reward.payoutEventType,
    ]
  }

  private func invokeError(_ error: NSError) {
    FlutterTapresearchPlugin.channel?.invokeMethod(
      "onError",
      arguments: ["code": String(error.code), "message": error.localizedDescription])
  }
}

File: flutter_tapresearch-1.4.0/android/src/main/kotlin/com/anavrinapps/flutter_tapresearch/FlutterTapresearchPlugin.kt
package com.anavrinapps.flutter_tapresearch

import android.app.Activity
import android.content.Context
import android.os.Handler
import android.os.Looper

import androidx.annotation.NonNull

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

// TapResearch SDK 3.x.
// NOTE: package paths below follow the 3.x docs; verify against the artifact
// you resolve (`com.tapresearch:tapsdk`) and adjust imports if they differ.
import com.tapresearch.tapsdk.TapResearch
import com.tapresearch.tapsdk.callback.TRContentCallback
import com.tapresearch.tapsdk.models.TRError
import com.tapresearch.tapsdk.models.TRReward

/**
 * Flutter bridge for the TapResearch SDK 3.x.
 *
 * The 3.x SDK requires both an api token and a user identifier at
 * initialization, uses string placement tags, and delivers rewards as a
 * batched list. See [TapResearch] (lib/flutter_tapresearch.dart) for the Dart
 * side of the protocol.
 */
class FlutterTapresearchPlugin : FlutterPlugin, MethodCallHandler, ActivityAware {

    private lateinit var channel: MethodChannel
    private var activity: Activity? = null
    private var appContext: Context? = null
    private val mainHandler = Handler(Looper.getMainLooper())

    override fun onAttachedToEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
        appContext = binding.applicationContext
        channel = MethodChannel(binding.binaryMessenger, "flutter_tapresearch")
        channel.setMethodCallHandler(this)
    }

    override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

    override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
        when (call.method) {
            "initialize" -> initialize(call, result)
            "isReady" -> result.success(TapResearch.isReady())
            "canShowContentForPlacement" -> {
                val tag = call.argument<String>("placementTag") ?: ""
                result.success(TapResearch.canShowContentForPlacement(tag) { error ->
                    invokeOnMain("onError", errorMap(error))
                })
            }
            "showContentForPlacement" -> showContent(call, result)
            "setUserIdentifier" -> {
                val userId = call.argument<String>("userIdentifier") ?: ""
                TapResearch.setUserIdentifier(userId) { error ->
                    invokeOnMain("onError", errorMap(error))
                }
                result.success(null)
            }
            "sendUserAttributes" -> {
                val attrs = call.argument<Map<String, Any>>("attributes") ?: emptyMap()
                TapResearch.sendUserAttributes(HashMap(attrs)) { error ->
                    invokeOnMain("onError", errorMap(error))
                }
                result.success(null)
            }
            else -> result.notImplemented()
        }
    }

    private fun initialize(call: MethodCall, result: Result) {
        val apiToken = call.argument<String>("apiToken")
        val userIdentifier = call.argument<String>("userIdentifier")
        val context = activity ?: appContext
        if (apiToken.isNullOrEmpty()) {
            result.error("no_api_token", "A null/empty api token was provided", null)
            return
        }
        if (userIdentifier.isNullOrEmpty()) {
            result.error("no_user_id", "A null/empty user identifier was provided", null)
            return
        }
        if (context == null) {
            result.error("no_context", "No Android context available for init", null)
            return
        }

        TapResearch.initialize(
            apiToken,
            userIdentifier,
            context,
            rewardCallback = { rewards -> invokeOnMain("onRewards", rewards.map(::rewardMap)) },
            errorCallback = { error -> invokeOnMain("onError", errorMap(error)) },
            sdkReadyCallback = { invokeOnMain("onSdkReady", null) }
        )
        result.success(null)
    }

    private fun showContent(call: MethodCall, result: Result) {
        val tag = call.argument<String>("placementTag") ?: ""
        val custom = call.argument<Map<String, Any>>("customParameters")
        val contentCallback = object : TRContentCallback {
            override fun onTapResearchContentShown(placementTag: String) {
                invokeOnMain("onContentShown", placementTag)
            }

            override fun onTapResearchContentDismissed(placementTag: String) {
                invokeOnMain("onContentDismissed", placementTag)
            }
        }
        TapResearch.showContentForPlacement(
            tag,
            if (custom != null) HashMap(custom) else null,
            contentCallback
        ) { error -> invokeOnMain("onError", errorMap(error)) }
        result.success(null)
    }

    private fun rewardMap(reward: TRReward): Map<String, Any?> = mapOf(
        "transactionIdentifier" to reward.transactionIdentifier,
        "placementIdentifier" to reward.placementIdentifier,
        "placementTag" to reward.placementTag,
        "currencyName" to reward.currencyName,
        "rewardAmount" to reward.rewardAmount,
        "payoutEventType" to reward.payoutEventType
    )

    private fun errorMap(error: TRError?): Map<String, Any?> = mapOf(
        "code" to (error?.code?.toString() ?: "unknown"),
        "message" to (error?.message ?: "")
    )

    /** MethodChannel callbacks must be delivered on the main thread. */
    private fun invokeOnMain(method: String, args: Any?) {
        mainHandler.post { channel.invokeMethod(method, args) }
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activity = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onDetachedFromActivity() {
        activity = null
    }
}
